api/indexer/indexer.go

```diff
@@ -8,230 +8,230 @@ import (
+	blk, ok := i.blockHeightToBlock[height]
 	if !ok {
 		return nil, fmt.Errorf("%w: height=%d", errBlockNotFound, height)
 	}
 	return blk, nil
 }
 
 func (i *Indexer) GetBlock(blkID ids.ID) (*chain.ExecutedBlock, error) {
-	height, ok := i.blockIDToHeight.Get(blkID)
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+
+	height, ok := i.blockIDToHeight[blkID]
 	if !ok {
 		return nil, fmt.Errorf("%w: %s", errBlockNotFound, blkID)
 	}
-	return i.GetBlockByHeight(height)
+	return i.getBlockByHeight(height)
 }
 
-func (i *Indexer) storeTransactions(blk *chain.ExecutedBlock) error {
-	batch := i.txDB.NewBatch()
-	defer batch.Reset()
-
-	for j, tx := range blk.Block.Txs {
-		result := blk.ExecutionResults.Results[j]
-		if err := i.storeTransaction(
-			batch,
-			tx.GetID(),
-			blk.Block.Tmstmp,
-			result.Success,
-			result.Units,
-			result.Fee,
-			result.Outputs,
-			string(result.Error),
-		); err != nil {
-			return err
-		}
-	}
-
-	return batch.Write()
-}
+func (i *Indexer) GetTransaction(txID ids.ID) (bool, *chain.Transaction, int64, *chain.Result, error) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
 
-func (*Indexer) storeTransaction(
-	batch database.KeyValueWriter,
-	txID ids.ID,
-	timestamp int64,
-	success bool,
-	units fees.Dimensions,
-	fee uint64,
-	outputs [][]byte,
-	errorStr string,
-) error {
-	storageTx := storageTx{
-		Timestamp: timestamp,
-		Success:   success,
-		Units:     units.Bytes(),
-		Fee:       fee,
-		Outputs:   outputs,
-		Error:     errorStr,
-	}
-	storageTxBytes := storageTx.MarshalCanoto()
-
-	return batch.Put(txID[:], storageTxBytes)
-}
-
-func (i *Indexer) GetTransaction(txID ids.ID) (bool, int64, bool, fees.Dimensions, uint64, [][]byte, string, error) {
-	v, err := i.txDB.Get(txID[:])
-	if errors.Is(err, database.ErrNotFound) {
-		return false, 0, false, fees.Dimensions{}, 0, nil, "", nil
-	}
-	if err != nil {
-		return false, 0, false, fees.Dimensions{}, 0, nil, "", err
+	cachedTx, ok := i.txCache[txID]
+	if !ok {
+		return false, nil, 0, nil, nil
 	}
-
-	storageTx := storageTx{}
-	if err := storageTx.UnmarshalCanoto(v); err != nil {
-		return false, 0, false, fees.Dimensions{}, 0, nil, "", fmt.Errorf("failed to unmarshal storage tx %s: %w", txID, err)
+	blk, ok := i.blockHeightToBlock[cachedTx.blkHeight]
+	if !ok {
+		return false, nil, 0, nil, nil
 	}
-
-	unpackedUnits, err := fees.UnpackDimensions(storageTx.Units)
-	if err != nil {
-		return false, 0, false, fees.Dimensions{}, 0, nil, "", fmt.Errorf("failed to unpack units for storage tx %s: %w", txID, err)
+	tx := blk.Block.Txs[cachedTx.index]
+	if len(blk.ExecutionResults.Results) <= cachedTx.index {
+		return false, nil, 0, nil, fmt.Errorf("%w: block height %d, transaction index %d", errTxResultNotFound, blk.Block.Hght, cachedTx.index)
 	}
-	return true, storageTx.Timestamp, storageTx.Success, unpackedUnits, storageTx.Fee, storageTx.Outputs, storageTx.Error, nil
+	result := blk.ExecutionResults.Results[cachedTx.index]
```

Comment:
Can we include index out of bounds checks for both tx and result?

Reply:
The revision is now returning the same error that mentions tx result not found in both cases ie. tx and result not being found both result in the error "transaction result not found", could we separate these two?

Reply:
np; please note that the case where the transaction index is missing is not possible, regardless of what input blocks are provided via Notify.